neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+***Description:***
+
+The function takes as input the list of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. The function calculates $\sum s_i \cdot p_i$.
+
+The multiplication on the scalar is the addition of that point a scalar number of times:
+
+$$
+s \cdot p = \underbrace{p + p + \ldots + p}_{s}
+$$
+
+The $E(F_p)$ curve,  points on the curve and the addition operation are defined in the BLS12-381 Curve Specification section.
+
+Note:
+
+- We take as input any points on the curve, not only from   $G_1$.
+- The scalar is an arbitrary unsigned integer and can be bigger than the group order.
+
+***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. Each point is encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and a scalar has a type `u256` and BigEndian encoded in 32 bytes. Expected `128*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the concatenation of uncompressed point from $E(F_p)$ — `96` bytes and scalar — `32` bytes. More details are in the Curve Points Encoding section.
+
+***Output:*** the output is 96 bytes — the one point $\in E(F_p)$ in decompressed form. For empty input it returns point on infinity (see  Curve Points Encoding section).
+
+***Gas Estimation:***
+
+This function should be calculated by Pippenger’s algorithm[^25]. The Complexity of this algorithm is $O(\frac{k}{\log(k)})$. For gas calculation we will use the formula $\frac{k}{\max(\log_2(k), 1)}$ the same way as in the host function for `BN254`[^10].
+
+```rust
+let k = (input_bytes+item_size-1)/item_size;
+let gas_consumed = A + B*k + C * if k > 1 {k / (k as f32).log2().floor()} else {k};
+```
+
+A, B and C are constants calculated empirically.
+
+For gas estimation, we can use the benchmark vectors for addition and multiplication for EIP-2537[^46].
+
+***Test cases:***
+
+The same test cases as for the bls12381_g1_sum section.
+
+Addition test cases:
+
+- `group_order * P = 0`
+- `(scalar + groupt_order) * P = scalar * P`
+- `P + P + P .. + P = N*P`
+- `0 * P = 0`
+- `1 * P = P`
+- Scalar is a MAX_INT
+
+***Tests References:***
+
+The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].
+
+***Error cases:***
+
+- The input length is not divided by 128
+- Too much memory is used
+- Field elements encoded incorrectly (see Curve points encoded section)
+- Any points not on the curve
+
+***Annotation:***
+
+```rust
+/// Computes multiexp on BLS12-381 curve using Pippenger's algorithm 
+///\sum_i si*pi should be equal result.
+///
+/// # Arguments
+///
+/// * `value` - sequence of (pi, si), where
+///    pi is point (x:Fp, y:Fp) on BLS12-381, and si is u256.
+///    BLS12-381 is Y^2 = X^3 + 4 curve over Fp.
+///
+///   `value` is encoded as packed, big-endian
+///   `[(([u8; 48], [u8; 48]), u256)]` slice.
```

Comment:
`value` is encoded as a concatenation of `(x, y, s)` tuples, where `x: [u8; 48]`, `y: [u8; 48]` and `s: [u8; 32]` encode a point and a scalar.

Reply:
@Ekleog-NEAR I'll make the explanation clearer

Reply:
just removed this part, because comments in the annotation don't add any new information: https://github.com/near/NEPs/pull/488/commits/cc448df84cb9d3a3d4dd73b14097b9ef011f2d6d